include/ginkgo/core/solver/chebyshev.hpp

```diff
@@ -0,0 +1,295 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_SOLVER_CHEBYSHEV_HPP_
+#define GKO_PUBLIC_CORE_SOLVER_CHEBYSHEV_HPP_
+
+
+#include <vector>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/combined.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+
+
+namespace gko {
+namespace solver {
+
+
+/**
+ * Chebyshev iteration is an iterative method that uses another coarse
+ * method to approximate the error of the current solution via the current
+ * residual. It has another term for the difference of solution. Moreover, this
+ * method requires knowledge about the spectrum of the matrix.
+ *
+ * ```
+ * solution = initial_guess
+ * while not converged:
+ *     residual = b - A solution
+ *     error = solver(A, residual)
+ *     solution = solution + alpha_i * error + beta_i * (solution_i -
+ * solution_{i-1})
```

Comment:
After looking a bit more into it, my proposal doesn't make too much sense. Perhaps instead it would be possible to extract the common residual computation.

Reply:
IR can add the moment parameter to have fixed beta * (sol_i - sol_{i-1}).
chebyshev alpha and beta are changed during iteration, so it is not quite suitable to put it in IR.
adding foci to IR with some flag may be able to switch the version.

Reply:
what I mean is that both have the following structure:
```
while not converged:
  residual = b - A * solution
  error = M^-1 * residual
  solution = update(residual, solution, ...)
```
And the only difference is `update(...)`. So perhaps both could use a unified implementation in that way.

Reply:
Yes, it's true and applies almost all iterative methods, I think.

Reply:
You're right. But what makes the Chebychev more similar to IR than the other solvers is the `if(iter == 0)` block. This is exactly the same, so perhaps this part could be extracted into a residual update function.

Reply:
Yes, I need to think how to extract them. It will contain many input parameters, is it sutiable?

Reply:
Can some of these parameters be passed via the workspace?

Reply:
`residual` is also part of the workspace, so my ideal would be using it like:
```
residual_update(solver, iter, dense_x, dense_b, criterion, log)
```
And everything else we can get from the workspace.

TBH, do we actually still need the `residual_ptr`? I think it's only relevant for the `iter == 0` branch, and otherwise its `residual_ptr = residual`. So we could just pass-in the initial guess mode.

Reply:
passing initial guess mode also takes one argument as passing `residual_ptr`, right?
I have thought about getting something from the workspace.
`create_...` is protected, so we need to use get_workspace_op with casting (need to cast const for residual) or lambda
It requires the vectors' names to be the same in IR/Chebyshev. It's true now, but we do not have any restrictions for that.
Thus, I think passing everything is simple